--- ngx_http_upstream.h.orig	2026-05-20 02:24:29.302902816 -0700
+++ ngx_http_upstream.h	2026-05-20 15:37:17.981957134 -0700
@@ -4,9 +4,10 @@
  * Copyright (C) Nginx, Inc.
  */
 
-
-#ifndef _NGX_HTTP_UPSTREAM_H_INCLUDED_
-#define _NGX_HTTP_UPSTREAM_H_INCLUDED_
+#ifndef _NGX_HTTP_SOCKS_UPSTREAM_H_INCLUDED_
+#define _NGX_HTTP_SOCKS_UPSTREAM_H_INCLUDED_
+//#ifndef _NGX_HTTP_UPSTREAM_H_INCLUDED_
+//#define _NGX_HTTP_UPSTREAM_H_INCLUDED_
 
 
 #include <ngx_config.h>
@@ -55,7 +56,8 @@
 #define NGX_HTTP_UPSTREAM_IGN_XA_CHARSET     0x00000100
 #define NGX_HTTP_UPSTREAM_IGN_VARY           0x00000200
 
-
+// Check if SOCKS proxy is not defined
+#ifndef SOCKS_PROXY
 typedef struct {
     ngx_uint_t                       status;
     ngx_msec_t                       response_time;
@@ -75,6 +77,7 @@
     ngx_array_t                      upstreams;
                                              /* ngx_http_upstream_srv_conf_t */
 } ngx_http_upstream_main_conf_t;
+#endif
 
 typedef struct ngx_http_upstream_srv_conf_s  ngx_http_upstream_srv_conf_t;
 
@@ -83,7 +86,8 @@
 typedef ngx_int_t (*ngx_http_upstream_init_peer_pt)(ngx_http_request_t *r,
     ngx_http_upstream_srv_conf_t *us);
 
-
+// Check if SOCKS proxy is not defined
+#ifndef SOCKS_PROXY
 typedef struct {
     ngx_http_upstream_init_pt        init_upstream;
     ngx_http_upstream_init_peer_pt   init;
@@ -112,7 +116,7 @@
     NGX_COMPAT_BEGIN(2)
     NGX_COMPAT_END
 } ngx_http_upstream_server_t;
-
+#endif
 
 #define NGX_HTTP_UPSTREAM_CREATE        0x0001
 #define NGX_HTTP_UPSTREAM_WEIGHT        0x0002
@@ -123,7 +127,8 @@
 #define NGX_HTTP_UPSTREAM_MODIFY        0x0040
 #define NGX_HTTP_UPSTREAM_MAX_CONNS     0x0100
 
-
+// Check if SOCKS proxy is not defined
+#ifndef SOCKS_PROXY
 struct ngx_http_upstream_srv_conf_s {
     ngx_http_upstream_peer_t         peer;
     void                           **srv_conf;
@@ -321,12 +326,13 @@
 
     ngx_resolver_ctx_t              *ctx;
 } ngx_http_upstream_resolved_t;
-
+#endif
 
 typedef void (*ngx_http_upstream_handler_pt)(ngx_http_request_t *r,
     ngx_http_upstream_t *u);
 
-
+// Check if SOCKS proxy is not defined
+#ifndef SOCKS_PROXY
 struct ngx_http_upstream_s {
     ngx_http_upstream_handler_pt     read_event_handler;
     ngx_http_upstream_handler_pt     write_event_handler;
@@ -423,10 +429,17 @@
     ngx_str_t   value;
     ngx_uint_t  skip_empty;
 } ngx_http_upstream_param_t;
-
+#endif
 
 ngx_int_t ngx_http_upstream_create(ngx_http_request_t *r);
-void ngx_http_upstream_init(ngx_http_request_t *r);
+// Check if SOCKS proxy is defined
+#ifdef SOCKS_PROXY
+	void ngx_http_socks_upstream_init(ngx_http_request_t *r);
+
+// Otherwise
+#else
+	void ngx_http_upstream_init(ngx_http_request_t *r);
+#endif
 ngx_int_t ngx_http_upstream_non_buffered_filter_init(void *data);
 ngx_int_t ngx_http_upstream_non_buffered_filter(void *data, ssize_t bytes);
 ngx_http_upstream_srv_conf_t *ngx_http_upstream_add(ngx_conf_t *cf,
